54/test algorithms tab

Painguin opened the pull request:
Tests the `Algorithms.js`, `AddAlgorithm.js` and `EditAlgorithm.js` pages.

close #54 .

pyphilia commented on test/algorithm.test.js:
```diff
@@ -0,0 +1,355 @@
+  REPLACEMENT_ALGORITHM,
+} from './fixtures/algorithms/algorithms';
+
+const menuGoTo = async (client, selector) => {
```
Place it in `utils.js`, and you could add a new `menuGoToAlgorithms` with the selector already set.

pyphilia commented on test/algorithm.test.js:
```diff
@@ -0,0 +1,355 @@
+  await el.click();
+};
+
+const verifyAlgorithm = async (client, algorithm) => {
```
Rename with something similar to `checkAlgorithmRowLayout`

pyphilia commented on test/algorithm.test.js:
```diff
@@ -0,0 +1,355 @@
+  { name, description, fileLocation },
+) => {
+  const fileOption = await client.$(`#${ADD_ALGORITHM_FROM_FILE_OPTION_ID}`);
+  fileOption.click();
```
`await` missing

pyphilia commented on test/algorithm.test.js:
```diff
@@ -0,0 +1,355 @@
+  });
+
+  it(
+    'Correctly adds Graasp algorithms',
```
`Correctly initialize Graasp algorithms`

pyphilia commented on test/algorithm.test.js:
```diff
@@ -0,0 +1,355 @@
+      const nAlgosPrior = await getNumberOfAlgorithms(client);
+      expect(nAlgosPrior > 0).to.be.true;
+
+      const deleteButton = await client.$(
```
It might be worth checking it is the right algorithm you deleted.

pyphilia commented on test/algorithm.test.js:
```diff
@@ -0,0 +1,355 @@
+  );
+
+  it(
+    'Adding a non existing algorithm does not add it to database',
```
I would go for `Using an incorrect algorithm file leads to error`

pyphilia commented on test/algorithm.test.js:
```diff
@@ -0,0 +1,355 @@
+  );
+
+  it(
+    'Adds, edits and deletes and algorithm',
```
typo `an algorithm`

pyphilia commented on test/fixtures/algorithms/algorithms.js:
```diff
@@ -0,0 +1,83 @@
+const DEFAULT_AUTHOR = 'Graasp';
+
+// eslint-disable-next-line import/prefer-default-export
+export const GRAASP_ALGORITHMS = [
```
I know you cannot access the constants because of `getPath`. But let's try to split the file to be able to get these constants.

pyphilia reviewed (changes requested):
Good job! What do you think of tests? :P If you have any feedback or comment about how we actually proceed let me know.

Maybe I would suggest to create a folder `algorithms` in `tests` and divide your big file into subfiles such as `editAlgorithm`, `addAlgorithm` as it will get even with our next features.. If a test does not belong to a single category let's also have a `common`/ `algorithm` file or something similar.

Painguin commented:
Thank you for your review @pyphilia ! 

I like how the tests are setup! It requires us to add a lot of ids and classes to our components only for the tests but I guess it is necessary :laughing:. (One alternative will be to use `client.react$` to access a component [(React Selectors)](https://webdriver.io/docs/selectors.html#react-selectors), but this requires spectron to [support the version 5 of WebdriverIO](https://stackoverflow.com/questions/56843461/how-to-select-a-react-component-with-spectron).
Otherwise I think tests are a good thing to ensure that future changes don't break previous ones.

As you suggested I split the algorithms tests into multiple test files.
I also changed the code when we set the sample database so that it also copies the graasp algorithm files again.

pyphilia reviewed (approved):
It would be really great to select react components directly!! But for now we have our ids and classes it's better than nothing x'3 
Looks good to me!!

pyphilia commented on public/app/config/graaspAlgorithms.js:
```diff
@@ -0,0 +1,73 @@
+    author: AUTHORS.GRAASP,
+    language: PROGRAMMING_LANGUAGES.PYTHON,
+  },
+  {
```
Just wondering, don't these algorithm need a type?

Painguin replied:
true :laughing:

pyphilia commented on public/app/db.js:
```diff
@@ -58,14 +58,11 @@ const bootstrapDatabase = (dbPath = DATABASE_PATH) => {
     // create the algorithms folder if it doesn't already exist
-    if (!fs.existsSync(algorithmsFolder)) {
-      await mkdirp(algorithmsFolder);
+    if (!fs.existsSync(ALGORITHMS_FOLDER)) {
```
I think you can  refactor it with `fs-extra` and `ensureDirSync` or something similar. There is an example in `test/application.js`

pyphilia commented on public/app/listeners/setSampleDatabase.js:
```diff
@@ -13,16 +14,21 @@ const {
   try {
     db.setState(_.cloneDeep(sampleDatabase)).write();
     fs.emptyDirSync(DATASETS_FOLDER);
+    fs.emptyDirSync(ALGORITHMS_FOLDER);
```
👍

pyphilia commented on test/algorithms/addAlgorithm.test.js:
```diff
@@ -0,0 +1,108 @@
+    mochaAsync(async () => {
+      app = await createApplication();
+      const { client } = app;
+      client.setTimeout({ implicit: 0 });
```
Interesting. What does this do?

pyphilia commented on test/algorithms/addAlgorithm.test.js:
```diff
@@ -0,0 +1,108 @@
+      app = await createApplication();
+      const { client } = app;
+      client.setTimeout({ implicit: 0 });
+      await menuGoToAlgorithms(client);
```
I can refactor this to include it in `commands` if you wish.

pyphilia commented on test/algorithms/addAlgorithm.test.js:
```diff
@@ -0,0 +1,108 @@
+    'Using an incorrect algorithm file location leads to error',
+    mochaAsync(async () => {
+      const { client } = app;
+      const nAlgosPrior = await getNumberOfAlgorithms(client);
```
Would rename as `nbAlgosPrev`

pyphilia commented on test/algorithms/addAlgorithm.test.js:
```diff
@@ -0,0 +1,108 @@
+      await addAlgorithmFromFileLocation(client, MISSING_FILE_ALGORITHM);
+      await clickAddAlgoSaveButton(client);
+
+      const nAlgosAfter = await getNumberOfAlgorithms(client);
```
Rename with `nbAlgosAfter`

pyphilia reviewed (changes requested):
Great! Here some comments.
